### PR TITLE
[Doc] [Tune] Fix tune resources user guide and faq to show how to define custom lambda to specify resources #17088

### LINKED ISSUE #17088: [tune] Clarify documentation around using different resource requirements across trials
[This discuss thread](https://discuss.ray.io/t/different-resource-amount-for-different-trials/509) shows how to run tune with different resource requirements for each trial. This is useful in settings where some models in the search require GPU and others do not. 

We can achieve this by taking advantage of the fact that the `resources_per_trial` are treated as part of the "hyperparameter space" under the hood:_
```
tune.run(
    ...
    config={
        ...
        "use_gpu": tune.grid_search([True, False])
    },
    resources_per_trial=tune.sample_from(
        lambda spec: {"gpu": 1} if spec.config.use_gpu else None
    )
)
```

The documentation should ideally address this behavior. It's super useful but hard to find!

### PR DESCRIPTION
**Pull request by heyitsmui**
## Why are these changes needed?

Users also wanted to know how to define custom lambda functions to specify resources per trial. We updated the tune resources user guide to show how to accomplish this using `tune.with_resources` or `ScalingConfig` and link them to relevant docs.

## Related issue number

Closes #17088

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by heyitsmui** [comment]
@krfricke @xwjiang2010 can you help me take a look when you get a chance? thanks!

File: doc/source/tune/faq.rst [hunk 1]
```diff
@@ -280,6 +280,33 @@ on other nodes as well. Please refer to the
 :ref:`placement groups documentation <ray-placement-group-doc-ref>` to learn more
 about these placement strategies.
 
+You can also allocate specific resources to a trial based on a custom rule via lambda functions.
+For instance, if you want to allocate GPU resources to trials based on a setting in your config:
+
+.. code-block:: python
+
+    tuner = tune.Tuner(
```
Comment (maxpumperla) [inline doc/source/tune/faq.rst:288]: @heyitsmui can you please make this a testable `literalinclude` block? That should be easy to convert, you can have a look at various examples in the same file.

**Review by maxpumperla — COMMENTED** [review]
@heyitsmui thanks, this is a great addition. Just one nit from my side.

File: doc/source/tune/faq.rst (hunk 1 above)
Comment (heyitsmui) [reply doc/source/tune/faq.rst:288]: thanks @maxpumperla! will add

Comment (heyitsmui) [reply doc/source/tune/faq.rst]: @maxpumperla just fixed, please take a look when you have a moment, thanks!

Comment (maxpumperla) [reply doc/source/tune/faq.rst]: @heyitsmui the linter fails now, other than that we should be good here. Fails in two places:

```
tuner = tune.Tuner(
--
  | tune.with_resources(
  | train_fn,
  | -            resources=lambda spec: {"GPU": 1} if config.use_gpu else {"GPU": 0})
  | +            resources=lambda spec: {"GPU": 1} if config.use_gpu else {"GPU": 0},
  | +        )
  | )
  | tuner.fit()
  | # __resources_lambda_end__
  | @@ -107,9 +108,11 @@ if not MOCK:
  | train_fn,
  | param_space={
  | "scaling_config": ScalingConfig(
  | -                trainer_resources=lambda spec: {"GPU": 1} if config.use_gpu else {"GPU": 0}
  | -                )
  | -        }
  | +                trainer_resources=lambda spec: {"GPU": 1}
  | +                if config.use_gpu
  | +                else {"GPU": 0}
  | +            )
  | +        },
  | )
```

Comment (heyitsmui) [reply doc/source/tune/faq.rst]: thanks! just put out a fix, will wait for all checks to pass first

Comment (heyitsmui) [reply doc/source/tune/faq.rst]: thanks again @maxpumperla all tests have passed now!

File: doc/source/tune/doc_code/faq.py [hunk 2]
```diff
@@ -93,6 +93,30 @@ def train(config):
     tuner.fit()
     # __resources_pgf_end__
 
+    # __resources_lambda_start__
+    tuner = tune.Tuner(
+        tune.with_resources(
+            train_fn,
+            resources=lambda spec: {"GPU": 1} if config.use_gpu else {"GPU": 0},
```
Comment (richardliaw) [inline doc/source/tune/doc_code/faq.py:100]: this should be spec.config.use_gpu? config is not defined here?

Comment (heyitsmui) [reply doc/source/tune/doc_code/faq.py:100]: @richardliaw good catch, let me just update spec to config to avoid confusion

**Review by Yard1 — APPROVED** [review]

**Review by xwjiang2010 — APPROVED** [review]
Thanks Michael!

**Review by krfricke — APPROVED** [review]
Thanks!

**Pull request merged**